crates/pipeline_manager/src/db/mod.rs

```diff
@@ -1161,7 +1285,50 @@ impl Storage for ProjectDB {
                 }
             })
             .collect();
-        Ok(vec)
+        Ok((tenant_id, vec))
+    }
+
+    async fn get_or_create_tenant_id(
+        &self,
+        tenant_name: String,
+        provider: String,
+    ) -> AnyResult<TenantId> {
+        let conn = self.pool.get().await?;
+        let res = conn
+            .query_opt(
+                "SELECT id FROM tenant WHERE tenant = $1 AND provider = $2",
+                &[&tenant_name, &provider],
+            )
+            .await?;
+        match res {
+            Some(row) => Ok(TenantId(row.get(0))),
+            None => {
+                self.create_tenant_if_not_exists(Uuid::now_v7(), tenant_name, provider)
+                    .await
+            }
+        }
+    }
+
+    async fn create_tenant_if_not_exists(
+        &self,
+        tenant_id: Uuid,
+        tenant_name: String,
+        provider: String,
+    ) -> AnyResult<TenantId> {
+        let conn = self.pool.get().await?;
+        // This part can have a race when there are two concurrent attempts
```

Comment:
I don't understand this race from the comment, why wouldn't the tenant_id be unique?

Reply:
The comment is slightly stale, so I'll fix it. Problem is that if `get_or_create_tenant_id` is called concurrently, `create_tenant_if_not_exists` will be called at the same time. Ideally, we'd have something like Java's `putIfAbsent` for maps, which inserts a new value if none exists, else gets the existing value. This single line INSERT is the best way I could find to mimic that. Splitting it into a separate INSERT followed by a SELECT is what's race prone (and that's what the comment was referring to).